src/eko/io/types.py

```diff
@@ -0,0 +1,162 @@
+"""Common type definitions, only used for static analysis.
+
+Unfortunately, Python has still some discrepancies between runtime classes and
+type hints, so it is better not to mix dataclasses and generics.
+
+E.g. before it was implemented::
+
+    @dataclasses.dataclass
+    class RunningReference(DictLike, Generic[Quantity]):
+        value: Quantity
+        scale: float
+
+but in this way it is not possible to determine that ``RunningReference`` is
+subclassing ``DictLike``, indeed::
+
+    inspec.isclass(RunningReference)  # False
+    issubclass(RunningReference, DictLike)  # raise an error, since
+                                            # RunningReference is not a class
+
+Essentially classes can be used for type hints, but types are not all classes,
+especially when they involve generics.
+
+For this reason I prefer the less elegant dynamic generation, that seems to
+preserve type hints.
+
+"""
+import dataclasses
+import enum
+import typing
+from math import isnan
+from typing import Any, Dict
+
+from .dictlike import DictLike
+
+T = typing.TypeVar("T")
+Quantity = typing.TypeVar("Quantity", bound=typing.Union[int, float])
+
+
+def reference_running(quantity: typing.Type[Quantity]):
+    """Generate running quantities reference point classes.
+
+    The motivation for dynamic generation is provided in module docstring.
+
```

Comment:
you should still provide a `Parameters` section (here and below)

Reply:
There are a lot of parameters/returns section missing, but they are not the priority. Moreover, it is better to have docstrings, especially for complex functions, but if the function is simple enough, with telling names for parameters and type hints, the parameters section is not that fundamental (i.e. I'd rather spend my time on something else).